Some unit tests fail on Windows

We need someone with a Windows machine who can fix and test this issue. Help would be greatly appreciated.

Notes on the various errors. Comments follow each error message. I'm not sure which one was fixed. 

**First Error**
```
testSortOn(tech.tablesaw.api.DateColumnTest)  Time elapsed: 0.003 sec  <<< ERROR!
java.lang.IllegalStateException: Column prefix: D is unknown.
	at tech.tablesaw.api.Table.sortOn(Table.java:484)
	at tech.tablesaw.api.DateColumnTest.testSortOn(DateColumnTest.java:154)
[...]
```
This error suggests that the column names were not read correctly.  The sort routine can't find a column that matches the one specified in the request:

        Table sorted = unsorted.sortOn("Date");
so it drops into the code that looks for a + or - sign prepended to the name. It gets the D from Date because there is no prepended sign. 

Looking at the test code, it's hard to see how the name could be read incorrectly since it's the first column and the delimiting comma follows immediately. 

```
        Table unsorted = Table.read().csv(
                "Date,1 Yr Treasury Rate" + System.lineSeparator()
```
A quick check showed no invisible characters in the header string.

**Second and Third Error**
```
testLocalDateDetectionEnglish(tech.tablesaw.io.csv.CsvReaderTest)  Time elapsed: 0.001 sec  <<< FAILURE!
java.lang.AssertionError: 
Expected: is <[LOCAL_DATE]>
     but: was <[]>
	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
	at org.junit.Assert.assertThat(Assert.java:956)
	at org.junit.Assert.assertThat(Assert.java:923)
	at tech.tablesaw.io.csv.CsvReaderTest.testLocalDateDetectionEnglish(CsvReaderTest.java:170)
[...]
```
```
testLocalDateDetectionFrench(tech.tablesaw.io.csv.CsvReaderTest)  Time elapsed: 0.003 sec  <<< FAILURE!
java.lang.AssertionError: 
Expected: is <[LOCAL_DATE]>
     but: was <[]>
	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
	at org.junit.Assert.assertThat(Assert.java:956)
	at org.junit.Assert.assertThat(Assert.java:923)
	at tech.tablesaw.io.csv.CsvReaderTest.testLocalDateDetectionFrench(CsvReaderTest.java:227)
[...]
```
These two errors have the same cause. In both cases we should see an array containing the columnType LOCAL_DATE, but the array is empy. (I think we'd see the word null printed if there was a null value. 
This seems like it might be related to the cyclic relationship issue, except that other manifestations throw an NPE reading the ColumnType array rather than completing and returning the wrong result. 

The code below suggests that can't be the issue, because the return value of detectedType is tested on the next line for equality so it isn't null. The method returns an array constructed on the last line so it executes to completion. Best guess is that the table is parsed incorrectly so that the entire thing is read as the header (which is skipped), no lines are passed to the type detector and the method returns without really doing anything. That gets us back to some kind of parsing error.

```
        for (List<String> valuesList : columnData) {
            ColumnType detectedType = detectType(valuesList, options);
            if (detectedType.equals(StringColumnType.STRING) && rowCount > STRING_COLUMN_ROW_COUNT_CUTOFF) {
                HashSet<String> unique = new HashSet<>(valuesList);
                double uniquePct = unique.size() / (valuesList.size() * 1.0);
                if (uniquePct > STRING_COLUMN_CUTOFF) {
                    detectedType = TEXT;
                }
            }
            columnTypes.add(detectedType);
        }
        return columnTypes.toArray(new ColumnType[0]);

```

**Fourth Error**
```
testReadFailure2(tech.tablesaw.io.csv.CsvReaderTest)  Time elapsed: 0.008 sec  <<< ERROR!
java.lang.IllegalArgumentException: Column Test is not present in table read_failure_test2.csv
	at tech.tablesaw.api.Table.columnIndex(Table.java:278)
	at tech.tablesaw.table.Relation.shortColumn(Relation.java:430)
	at tech.tablesaw.io.csv.CsvReaderTest.testReadFailure2(CsvReaderTest.java:392)
[...]
```
This has the same underlying cause as the first error, I think. This time the code reads the table from a file with comma delimiters separating columns, but somehow it fails to create the first column. Could we be getting errors on windows reading commas?  

**Fifth Error**
```
test_JavaBehaviour(tech.tablesaw.io.csv.UnicodeBomHandlingTest)  Time elapsed: 0 sec  <<< FAILURE!
java.lang.AssertionError: expected:<1> but was:<ï»¿1>
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotEquals(Assert.java:834)
	at org.junit.Assert.assertEquals(Assert.java:118)
	at org.junit.Assert.assertEquals(Assert.java:144)
	at tech.tablesaw.io.csv.UnicodeBomHandlingTest.test_JavaBehaviour(UnicodeBomHandlingTest.java:34)
[...]
```
This is a unicode bom error, so it's a different Windows disease from the line ending issue.

**Sixth Error**
```
html(tech.tablesaw.io.DataFrameWriterTest)  Time elapsed: 0 sec  <<< FAILURE!
org.junit.ComparisonFailure: expected:<...></td></tr>
</tbody>[]
> but was:<...></td></tr>
]
>
	at org.junit.Assert.assertEquals(Assert.java:115)
	at org.junit.Assert.assertEquals(Assert.java:144)
	at tech.tablesaw.io.DataFrameWriterTest.html(DataFrameWriterTest.java:58)
```
I can't tell which test is failing, and the expected results don't look like they could be produced by either test.

--- a/core/src/main/java/tech/tablesaw/io/Source.java
+++ b/core/src/main/java/tech/tablesaw/io/Source.java
@@ -15,16 +15,14 @@ import java.util.Scanner;
 
 public class Source {
 
+    // we always have one of these (file, reader, or inputStream)
     protected final File file;
     protected final Reader reader;
     protected final InputStream inputStream;
     protected final Charset charset;
 
     public Source(File file) {
-        this.file = file;
-        this.reader = null;
-        this.inputStream = null;
-        this.charset = Charset.defaultCharset();
+        this(file, Charset.defaultCharset());
     }
 
     public Source(File file, Charset charset) {
@@ -42,10 +40,7 @@ public class Source {
     }
 
     public Source(InputStream inputStream) {
-        this.file = null;
-        this.reader = null;
-        this.inputStream = inputStream;
-        this.charset = Charset.defaultCharset();
+        this(inputStream, Charset.defaultCharset());
     }
 
     public Source(InputStream inputStream, Charset charset) {

--- a/core/src/main/java/tech/tablesaw/io/fixed/FixedWidthReadOptions.java
+++ b/core/src/main/java/tech/tablesaw/io/fixed/FixedWidthReadOptions.java
@@ -29,7 +29,6 @@ import java.util.Locale;
 
 public class FixedWidthReadOptions extends ReadOptions {
 
-    // we always have one of these (file, reader, or inputStream)
     private final ColumnType[] columnTypes;
     private final FixedWidthFields columnSpecs;
     private final String lineEnding;
@@ -172,6 +171,10 @@ public class FixedWidthReadOptions extends ReadOptions {
         public Builder lineEnding(String lineEnding) {
             this.lineEnding = lineEnding;
             return this;
+        }
+
+        public Builder systemLineEnding() {
+            return lineEnding(System.lineSeparator());
         }
 
         public Builder padding(char padding) {

--- a/core/src/test/java/tech/tablesaw/io/csv/CsvWriterTest.java
+++ b/core/src/test/java/tech/tablesaw/io/csv/CsvWriterTest.java
@@ -21,7 +21,7 @@ public class CsvWriterTest {
         Table table = Table.create("testTable", colA, colB);
         StringWriter writer = new StringWriter();
         table.write().toWriter(writer, "csv");
-        assertEquals("colA,colB\na,1\nb,2\n", writer.toString());
+        assertEquals("colA,colB\na,1\nb,2\n", writer.toString().replaceAll("\\r\\n", "\n"));
     }
 
 }

--- a/core/src/test/java/tech/tablesaw/io/fixed/FixedWidthReaderTest.java
+++ b/core/src/test/java/tech/tablesaw/io/fixed/FixedWidthReaderTest.java
@@ -53,7 +53,7 @@ public class FixedWidthReaderTest {
                 .columnTypes(car_types)
                 .columnSpecs(car_fields_specs)
                 .padding('_')
-                .lineEnding("\n")
+                .systemLineEnding()
                 .build()
         );
 
@@ -75,7 +75,7 @@ public class FixedWidthReaderTest {
                 .columnTypes(car_types_with_SKIP)
                 .columnSpecs(car_fields_specs)
                 .padding('_')
-                .lineEnding("\n")
+                .systemLineEnding()
                 .build());
 
         assertEquals(4, table.columnCount());
@@ -92,7 +92,7 @@ public class FixedWidthReaderTest {
                 .columnTypes(car_types_with_SKIP)
                 .columnSpecs(car_fields_specs)
                 .padding('_')
-                .lineEnding("\n")
+                .systemLineEnding()
                 .skipTrailingCharsUntilNewline(true)
                 .build());
 
@@ -110,7 +110,7 @@ public class FixedWidthReaderTest {
                 .header(true)
                 .columnSpecs(car_fields_specs)
                 .padding('_')
-                .lineEnding("\n")
+                .systemLineEnding()
                 .sample(false)
                 .locale(Locale.getDefault())
                 .minimizeColumnSizes()
@@ -129,7 +129,7 @@ public class FixedWidthReaderTest {
                 .header(true)
                 .columnSpecs(car_fields_specs)
                 .padding('_')
-                .lineEnding("\n")
+                .systemLineEnding()
                 .missingValueIndicator("null")
                 .minimizeColumnSizes()
                 .sample(false)
@@ -153,7 +153,7 @@ public class FixedWidthReaderTest {
                 .columnTypes(car_types)
                 .columnSpecs(car_fields_specs)
                 .padding('_')
-                .lineEnding("\n")
+                .systemLineEnding()
                 .skipTrailingCharsUntilNewline(true)
                 .build()
         );

